core/reactor.h

```diff
@@ -106,6 +106,34 @@ do { \
     _lf_set_present(&out->is_present); \
 } while(0)
 
+/**
+ * Set the specified output (or input of a contained reactor)
+ * to the specified dynamic value and frees the value if no
+ * port references it.
+ *
+ * This version is used for any dynamically allocated types.
+ * The *reference* to "val" is copied (shallow copy) to "out". 
+ *
+ * At the beginning of each timestep, if no port 
+ * references "val", the runtime will free "val".
+ * 
+ * If "destructor" is not NULL, destructor(val) is called.
+ * Otherwise, free(val) is called.
+ * 
+ * @param out The output port (by name) or input of a contained
+ *  reactor in form input_name.port_name.
+ * @param val The value to insert into the self struct.
+ * @param dtor The function pointer used to free "val" in
+ *             the form of destructor(val).
+ *             If NULL, free(val) is used instead.
+ */
+#define _LF_SET_DYNAMIC(out, val, dtor) \
```

Comment:
The reason I used dtor instead of destructor is that C macros does something weird if the parameter name is the same as the field name when I do `token->destructor = destructor;`. It seems to throw an error during compile time.